### PR TITLE
Emit stackalloc block initializer for byte-sized enums

### PR DESCRIPTION
**Pull request by alrz**

### CODE REVIEW

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ private void EmitStackAllocInitializers(TypeSymbol type, BoundArrayInitializatio
                         EmitElementStackAllocInitializers(elementType, initExprs, includeConstants: false);
                     }
                 }
-                else if (elementType.SpecialType.SizeInBytes() == 1)
+                else if (elementType.EnumUnderlyingTypeOrSelf().SpecialType.SizeInBytes() == 1)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs:50]: Not related to this PR: we could also lift the 1 byte restriction as was done in the [CreateSpan PR](https://github.com/dotnet/roslyn/pull/61414). Somehow I didn't realize until now.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStackAllocInitializerTests.cs [hunk 2]
```diff
@@ -583,6 +583,63 @@ .maxstack  4
...
+        Print(p1);
+        Print(p2);
+    }
+
+    enum E : byte { A, B, C, D }
+}
+""";
+            CompileAndVerify(text,
+                options: TestOptions.UnsafeReleaseExe,
+                verify: Verification.Fails, expectedOutput: @"ABCDDD").VerifyIL("C.Main",
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStackAllocInitializerTests.cs:614]:
```suggestion
                verify: Verification.Fails, expectedOutput: @"ABCDDD").VerifyIL("C.Main",
```

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStackAllocInitializerTests.cs [hunk 3]
```diff
@@ -583,6 +583,63 @@ .maxstack  4
...
+  IL_0014:  ldc.i4.3
+  IL_0015:  initblk
+  IL_0017:  ldloc.0
+  IL_0018:  call       "void C.Print(C.E*)"
+  IL_001d:  call       "void C.Print(C.E*)"
+  IL_0022:  ret
+}
+""");
+        }
+
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenStackAllocInitializerTests.cs:642]:
```suggestion

```

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs (hunk 1 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs:50]: https://github.com/dotnet/roslyn/pull/61414 doesn't do anything on stackalloc, unlike https://github.com/dotnet/roslyn/pull/57123 but I think that's not merged in.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs:50]: Right, I mean we can do something like https://github.com/dotnet/roslyn/pull/61414 but for stackalloc. I didn't realize we'd already done a demo branch with exactly that. Thanks for digging up https://github.com/dotnet/roslyn/pull/57123. I'll ping there to see whether we should make a real PR with that change.

Comment (stephentoub) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs:50]: @jcouv, what became of this? Even just emitting it as a copy from the data section (which the JIT should be able to unroll in many cases) would likely be useful.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs:50]: @stephentoub It looks like the hackathon project didn't transfer into a committed project/PR. The [last comment from Jared there](https://github.com/dotnet/roslyn/pull/57123#issuecomment-1108741115) is that this just needs keyboard time, but we support it in principle. 
We're out of dev bandwidth for .NET 8 (ie. at least for one month) unless urgent. Let's discuss around that time. We could file an 17.9 issue (refresh, resolve conflicts and bring https://github.com/dotnet/roslyn/pull/57123 into `main`) for tracking purpose.

Comment (stephentoub) [reply src/Compilers/CSharp/Portable/CodeGen/EmitStackAllocInitializer.cs:50]: I opened https://github.com/dotnet/roslyn/issues/69325. Thanks.

**Pull request merged**